eclipse-theia/theia
# Cache instances from call hierarchy provider

#### What it does


As part of #10310, it was found that Theia does not satisfy the expectations of VSCode CLangD (or any other language client based on `vscode-languageclient`) because we do not guarantee that the plugin will receive the same instances of `CallHierarchyItem`s that it creates when calls are made.

This PR addresses that problem by caching call hierarchy items received from plugins and retrieving them when passing them back into the plugin.

In addition, it moves some way to addressing #10528, but with a twist. Since call hierarchy items are now being cached, there's need for a signal to clear the cache, and the return from `CallHierarchyAdapter.prepareCallHierarchy` is now a `CallHierarchySession` that includes a disposable to dispose of the cache. That isn't part of the API available to plugins, so the VSCode references view never disposes of its sessions (as far as I can tell, `CallHierarchySession`s are never disposed of in VSCode), but our internal view _does_ know when to dispose of its sessions, so I have left the view in, but adjusted the types to eliminate our purely internal types in favor of close mirrors of the plugin types.

#### How to test


1. Install the latest CLangD plugin
> To install the CLangD plugin, add this line to the `.theiaPlugins` field of the repo's `package.json`:
```
"vscode-clangd": "https://open-vsx.org/api/llvm-vs-code-extensions/vscode-clangd/0.1.13/file/llvm-vs-code-extensions.vscode-clangd-0.1.13.vsix"
```
2. Open the `examples/cpp-debug-workspace` from the [Theia CPP Extensions](https://github.com/eclipse-theia/theia-cpp-extensions) repo
3. Open `a.cpp`
4. Use `Calls: Show Call Hierarchy` on one of the functions there other than `main`.
5. You should get results, and you should be able to expand those results.
6. Call hierarchy for languages like TS, etc. should continue to work as before in both references view and internal view.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
